google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/TopicBacklogReaderImpl.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.pubsublite.internal;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutures;
+import com.google.cloud.pubsublite.Offset;
+import com.google.cloud.pubsublite.Partition;
+import com.google.cloud.pubsublite.TopicPath;
+import com.google.cloud.pubsublite.proto.ComputeMessageStatsResponse;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.protobuf.Timestamp;
+import java.time.Instant;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public final class TopicBacklogReaderImpl implements TopicBacklogReader {
+  private final TopicStatsClient client;
+  private final TopicPath topicPath;
+
+  public TopicBacklogReaderImpl(TopicStatsClient client, TopicPath topicPath) {
+    this.client = client;
+    this.topicPath = topicPath;
+  }
+
+  private static Optional<Instant> minTimestamp(Optional<Instant> t1, Instant t2) {
+    if (!t1.isPresent() || t2.isBefore(t1.get())) {
+      return Optional.of(t2);
+    }
+    return t1;
+  }
+
+  private static Instant toJavaInstant(Timestamp timestamp) {
+    return Instant.ofEpochSecond(timestamp.getSeconds(), timestamp.getNanos());
+  }
+
+  private static Timestamp toProtoTimestamp(Instant instant) {
+    return Timestamp.newBuilder()
+        .setSeconds(instant.getEpochSecond())
+        .setNanos(instant.getNano())
+        .build();
+  }
+
+  @Override
+  public ApiFuture<ComputeMessageStatsResponse> computeMessageStats(
+      Map<Partition, Offset> subscriptionState) {
+    List<ApiFuture<ComputeMessageStatsResponse>> perPartitionFutures =
+        subscriptionState.entrySet().stream()
+            .map(
+                e ->
+                    client.computeMessageStats(
+                        topicPath, e.getKey(), e.getValue(), Offset.of(Integer.MAX_VALUE)))
+            .collect(Collectors.toList());
+    return ApiFutures.transformAsync(
+        ApiFutures.allAsList(perPartitionFutures),
+        responses -> {
+          Optional<Instant> minPublishTime = Optional.empty();
```

Comment:
You don't have to keep shuffling back and forth between instant and timestamp. Instead, use the protobuf Timestamps.compare method.

Reply:
Ah, cool, done